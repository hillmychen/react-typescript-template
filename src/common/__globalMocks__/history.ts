/*
 * @Author: Hughie
 * @Date: 2020-06-18 19:17:50
 * @LastEditTime: 2020-06-20 13:54:56
 * @LastEditors: Hughie
 * @Description: 用于mock history，主要是测试组件的各种跳转
 */
import { createMemoryHistory } from 'history'

export const history = createMemoryHistory()
export default history
