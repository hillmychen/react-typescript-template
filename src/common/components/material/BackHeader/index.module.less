@import '~variables';

.back {
  display: flex;
  align-items: center;
  flex-shrink: 0;
  height: 50px;
  font-size: 16px;
  font-weight: bolder;
  background: #fff;
  box-shadow: @box-shadow;
  z-index: 1;
  & > span {
    cursor: pointer;
    height: 50px;
    line-height: 50px;
    padding: 0 15px;
    .backIcon {
      color: var(--primary-color);
      font-size: 18px;
    }
    .backText {
      margin-left: 5px;
    }
  }
}

.disabled {
  i {
    color: var(--disabled-color);
  }
  color: var(--disabled-color);
  & > span {
    cursor: not-allowed;
  }
}
