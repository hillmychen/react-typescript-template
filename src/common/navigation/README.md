# Navigation

> 扩展 `history` 路由跳转的功能，整个应用关于导航相关的操作都应该使用 `navigation`

#### 路由跳转案例

```ts
// 直接跳转
navigation.push('/url')

// 跳转携带 state
navigation.push('/url', { name: 'jack' })

// 跳转携带 query, query 对象会生成 search 字符串，替换原本的 search 字段
history.push({
  pathname: '/url',
  search: '?name=jack&age=20',
})
// vs
navigation.push({
  pathname: 'url',
  query: {
    name: 'jack',
    age: 20,
  },
})

// 同上
navigation.replace({
  pathname: 'url',
  query: {
    name: 'jack',
  },
})

// 前进后退
navigation.go(-1)
navigation.goBack()
navigation.goForward()

// 页面跳转保存数据，详情查看下面章节
navigation.pushKeepCache('/url', query)
```

#### 跳转保存当前路由数据数据

```ts
// 1. 组件加载时获取缓存的数据
const cache = navigation.cache.popItem<Cache>(cacheKey)

// 2. 页面跳转之前存储数据
navigation.cache.setItem(cacheKey, data)
navigation.pushKeepCache('/url', query)

// 3. 组件卸载清除
navigation.cache.removeItem(cacheKey)
```
