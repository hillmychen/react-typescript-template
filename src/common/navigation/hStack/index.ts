/*
 * @Author: Hughie
 * @Date: 2019-11-20 13:35:47
 * @Last Modified by: Hughie
 * @Last Modified time: 2019-11-21 09:28:55
 * 用于状态保存
 */

export * from './HStack'
