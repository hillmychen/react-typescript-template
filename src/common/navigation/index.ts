export * from './navigation'
export * from './utils'
export * from './history'
