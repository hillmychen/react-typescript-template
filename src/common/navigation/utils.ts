import querystring from 'query-string'
import { history } from './history'

/**
 * 获取 query 的值
 */
export function getQuery<T = any>(): Partial<T> {
  return (querystring.parse(history.location.search) as unknown) as Partial<T>
}

/**
 * 生成 search 字符串
 * @param searchObject json 对象
 * @param withCurrentSearch 是否将当前 search 携带上
 */
export function createSearch(searchObject: AnyObject, withCurrentSearch?: boolean) {
  const finalLocation = {
    ...searchObject,
    ...(withCurrentSearch ? querystring.parse(history.location.search) : null),
  }
  const searchString = querystring.stringify(finalLocation)
  return searchString ? `?${searchString}` : searchString
}
