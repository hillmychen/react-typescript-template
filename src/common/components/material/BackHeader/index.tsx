import React from 'react'
import { Breadcrumb } from 'antd'
import { Icon } from '../../basic/Icon'
import { history } from '@/common/navigation'
import styles from './index.module.less'
import cx from 'classnames'
import { useDidUpdate, useDidMount } from '@/common/hooks/lifecycle'
import { useSetState } from '@/common/hooks/utils/useSetState'

export interface BackHeaderProps {
  backText?: string | Array<string>
  disabled?: boolean
}
export interface BackHeaderState {
  texts: Array<string>
}

export const BackHeader: React.FC<BackHeaderProps> = props => {
  const [state, setState] = useSetState<BackHeaderState>({
    texts: [],
  })
  const updateText = () => {
    if (props?.backText) {
      let temp: Array<string> = []
      if (typeof props?.backText === 'string') {
        temp = props?.backText?.split('/')
      }
      if (Array.isArray(props?.backText)) {
        temp = Array.from(props?.backText)
      }
      setState({
        texts: temp,
      })
    }
  }
  // 禁止返回
  const handleGoback = () => {
    if (props.disabled) return
    history.goBack()
  }
  useDidMount(() => {
    updateText()
  })

  useDidUpdate(() => {
    updateText()
  }, [props.backText])
  return (
    <div className={cx(styles.back, { [styles.disabled]: history.length <= 1 })}>
      {!props.disabled && (
        <span data-testid="click" onClick={handleGoback} style={{ cursor: 'pointer' }}>
          <Icon type="left" className={styles.backIcon} />
        </span>
      )}
      {state.texts && state.texts.length > 0 && (
        <Breadcrumb>
          {state.texts.map(_ => (
            <Breadcrumb.Item key={_}>
              <span className={styles.backText}>{_}</span>
            </Breadcrumb.Item>
          ))}
        </Breadcrumb>
      )}
      {props.children}
    </div>
  )
}

export default BackHeader
