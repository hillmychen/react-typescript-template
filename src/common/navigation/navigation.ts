import { hStack } from './hStack'
import { history } from './history'
import { createSearch } from './utils'
import type { LocationDescriptorObject, Path, LocationState } from 'history'

export interface NavigationLocation extends Omit<LocationDescriptorObject, 'search'> {
  query?: string | AnyObject
}

class Navigation {
  readonly history = history

  /** 路由缓存 */
  get cache() {
    return hStack.contextCache
  }

  private transformLocation(location: NavigationLocation) {
    return { ...location, ...(typeof location.query === 'object' ? { search: createSearch(location.query) } : null) } as LocationDescriptorObject
  }

  init() {
    hStack.hookHistory(history)
  }

  /** 保持 Navigation.cache 不被清除的，并且跳转页面 */
  pushKeepCache(url: string, query?: AnyObject) {
    hStack.push(url, query)
  }

  /**
   * 同 history.push，search 替换为 query，支持对象
   * @example history.push({ pathname: '/url', query: { id: 'xx' } })
   */
  push(location: NavigationLocation): void

  push(path: Path, state?: LocationState): void

  push(location: Path | NavigationLocation, state?: LocationState) {
    if (typeof location === 'string') {
      this.history.push(location, state)
      return
    }
    this.history.push(this.transformLocation(location))
  }

  /**
   * 同 history.replace，search 替换为 query，支持对象
   * @example history.replace({ pathname: '/url', query: { id: 'xx' } })
   */
  replace(location: NavigationLocation): void

  replace(path: Path, state?: LocationState): void

  replace(location: Path | NavigationLocation, state?: LocationState) {
    if (typeof location === 'string') {
      this.history.replace(location, state)
      return
    }
    this.history.replace(this.transformLocation(location))
  }

  goBack = () => {
    this.history.goBack()
  }

  goForward = () => {
    this.history.goForward()
  }

  go = (n: number) => {
    this.history.go(n)
  }

  /**
   * 导航到登录页面
   * @description 使用 location.href 跳转到登录页，这样会走 nginx 服务器重定向
   */
  toLogin() {
    window.location.href = '/login'
  }

  /** 打开新的标签页，且开启 noopener， noreferrer */
  openNewWindow(url: string) {
    const a = document.createElement('a')
    a.href = url
    a.rel = 'noopener noreferrer'
    a.target = '_blank'
    document.body.appendChild(a)
    a.click()
    a.remove()
  }
}

export const navigation = new Navigation()
