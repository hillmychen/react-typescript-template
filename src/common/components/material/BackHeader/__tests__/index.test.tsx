import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import BackHeader from '../index'
import userEvent from '@testing-library/user-event'
import { history } from '@/common/navigation/history'

jest.mock('@/common/navigation/history', () => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  return require('@/common/__globalMocks__/history')
})

beforeAll(() => {
  history.push('/a')
})

it('BackHeader snapshot', () => {
  const { asFragment } = render(<BackHeader backText="返回" />)
  expect(asFragment()).toMatchSnapshot()
})

it('BackHeader onClick', () => {
  history.goBack = jest.fn()
  const { getByTestId } = render(<BackHeader backText="返回" />)
  const elem = getByTestId('click')
  expect(elem).not.toBeUndefined()
  userEvent.click(elem)
  expect(history.goBack).toHaveBeenCalled()
})
