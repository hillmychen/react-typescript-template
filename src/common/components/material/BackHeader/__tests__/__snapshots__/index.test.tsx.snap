// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackHeader snapshot 1`] = `
<DocumentFragment>
  <div
    class="back"
  >
    <span
      data-testid="click"
    >
      <span
        class="anticon backIcon"
        role="img"
      >
        <svg
          aria-hidden="true"
          class=""
          fill="currentColor"
          focusable="false"
          height="1em"
          width="1em"
        >
          <use
            xlink:href="#alu-icon-fanhui"
          />
        </svg>
      </span>
       
      <span
        class="backText"
      >
        返回
      </span>
    </span>
  </div>
</DocumentFragment>
`;
